### ISSUE TITLE
Review, test, integrate and report on the multiple options for Networking with Kubernetes (for the Kubernetes Runner)

### ISSUE THREAD

**Issue by daviddias** [ipfs/testground#299]
After testing and verifying that the k8s runner enables us so spawn multiple thousands of containers (great work @nonsense !) Our crux is to get a networking layer that enables us to have the control network (aka control plane) and the data network (aka data plane) that enables us to configure the network (set IP addresses, iptables/firewalls, latency, jitter, bandwidth, etc).

**Comment by nonsense** [comment]
@daviddias thanks for the task. We should probably evaluate those in two contexts: various CNIs under managed Kubernetes environments - AWS EKS, DO, GKE ; as well as under self maintained Kubernetes clusters.

There are pros/cons to both approaches, but given the time requirements, I suggest we start with understanding how various CNIs fit with GKE, DO and EKS. Does that make sense and do you agree?

**Comment by daviddias** [comment]
@nonsense it does indeed! Wanna create a markdown table as a grid and fill it as you progress with the testing?

**Comment by nonsense** [comment]
Here is an overview of what we tried on managed Kubernetes:

1. Digital Ocean (which run Cilium CNI) - we tried running `Multus CNI` with `Flannel CNI` on it. We managed to get a second interface on a pod, but both interfaces were picking up IPs from the Cilium network, and not from the Flannel network, even though they were annotated with the Flannel config.

Ultimately I am not convinced that `Multus CNI` works with Cilium (and the Digital Ocean Kubernetes implementation), but will investigate further in case we did something wrong with the configuration.

2. AWS EKS (which runs AWS VPC CNI) - we tried to remove the `aws-node` which runs the `Amazon VPC CNI`  and install `Flannel` on it, but this failed since AWS doesn't allow us to set pod CIDR via the API (https://medium.com/@jeremy.i.cowan/running-flannel-on-eks-9a2f7a285a23). Will investigate further if it is possible to run AWS EKS with a custom CNI, but this seems quite hard, and there are not many docs that mention any solution. We want to not run AWS VPC CNI at all, because it is slow, but even if we run it, then it is not clear how we can have another CNI alongside it, so that we have a second overlay network.

3. Google's GKE (which runs GKE CNI by default) - we couldn't install Multus CNI at all on it.

---

Bottom line - at the moment we are failing to start a second overlay network alongside any Kubernetes default cluster network when we use a managed offering (AWS, GKE, or DO).

Given the amount of docs online about Flannel + Multus, it will probably be possible to run multiple networks on Kubernetes, if we run Kubernetes from scratch on a given cloud, and don't use managed offerings, but this is adding yet more complexity - after all we want this to be a self-serviced solution that developers can quickly use.

---

It is worth investigation GKE + Istio if it fits our use-case, because Istio seems to be supported on GKE. Most of our testing today was focused on DO + Multus + Flannel, so we'll continue to look into this.

**Comment by raulk** [comment]
Istio/Envoy is made for microservice routing, and policy enforcement. It is a poor fit for all of this.

> if we run Kubernetes from scratch on a given cloud, and don't use managed offerings, but this is adding yet more complexity - after all we want this to be a self-serviced solution that developers can quickly use.

As I’ve said before several times, the reason we want to own the K8S deployment is to not tie ourselves to proprietary clouds, and to have the degree of configurability we desire. Like precisely the one we are needing right now, and not finding with OOTB cloud solutions that are meant to run microservices, not ephemeral test workloads that require traffic shaping and other unusual low-level tweaks.

Self-service UX doesn’t imply vendor lock-in. We can make self-service solutions ourselves by automating the K8S cluster creation with CloudFormation and a bunch of AMIs, or with a script that runs the terraform and ansible formulae we already have.

**Comment by nonsense** [comment]
> Self-service UX doesn’t imply vendor lock-in. We can make self-service solutions ourselves by automating the K8S cluster creation with CloudFormation and a bunch of AMIs, or with a script that runs the terraform and ansible formulae we already have.

Having all this in a week or two is close to impossible for such a scale. It should also be easy for developers to setup and it should be easy to setup.

I completely agree this is the long term solution, but under the deadlines we are, and this ticket is part of that deadline as far as I understand (@daviddias) I cannot commit to implementing custom:
1. Ansible playbooks
2. Terraform playbooks
3. Packer configuration
4. Kubernetes configuration
5. K8s runner Go code
6. Docker registry so that we don't incur costs.

Let's stop going in circles on self-service vs managed vs vendor lock-in, when we are talking about MVPs, or let's adjust our deadlines or expectations.

**Comment by raulk** [comment]
Given we have a path forward, and we've implemented it already, I consider this epic to have concluded. I've moved the Calico, NetworkServiceMesh and cilium.io to Icebox state, since presumably won't be investigating those further if everything plays out correctly with our setup.
